internal/worker/httpserverargs/worker.go

```diff
@@ -46,34 +51,38 @@ func (w workerConfig) Validate() error {
 }
 
 type argsWorker struct {
-	catacomb                catacomb.Catacomb
-	cfg                     workerConfig
-	authenticator           macaroon.LocalMacaroonAuthenticator
-	managedCtrlConfigGetter *managedCtrlConfigGetter
+	catacomb        catacomb.Catacomb
+	cfg             workerConfig
+	authenticator   macaroon.LocalMacaroonAuthenticator
+	managedServices *managedServices
 }
 
 func newWorker(cfg workerConfig) (worker.Worker, error) {
 	if err := cfg.Validate(); err != nil {
 		return nil, errors.Trace(err)
 	}
 	w := argsWorker{
-		cfg:                     cfg,
-		managedCtrlConfigGetter: newManagedCtrlConfigGetter(cfg.controllerConfigGetter),
+		cfg: cfg,
+		managedServices: newManagedServices(
+			cfg.controllerConfigService,
+			cfg.userService,
+		),
 	}
 
 	if err := catacomb.Invoke(catacomb.Plan{
 		Site: &w.catacomb,
 		Work: w.loop,
 		Init: []worker.Worker{
-			w.managedCtrlConfigGetter,
+			w.managedServices,
 		},
 	}); err != nil {
 		return nil, errors.Trace(err)
 	}
 
 	authenticator, err := w.cfg.newStateAuthenticatorFn(
 		w.cfg.statePool,
-		w.managedCtrlConfigGetter,
+		w.managedServices,
```

Comment:
I can see that by combining the upstream services, you save a Goroutine, but this looks a bit weird. Not sure how I feel about it - the only alternative is a tomb per service.

Reply:
The advantage of one tomb for all services is that we can ensure the all get closed at the same time.